java-checks/src/main/java/org/sonar/java/checks/aws/AwsRegionShouldBeSetExplicitlyCheck.java

```diff
@@ -0,0 +1,217 @@
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+
+@Rule(key = "S6241")
+public class AwsRegionShouldBeSetExplicitlyCheck extends IssuableSubscriptionVisitor {
+
+  private static final String SDK_CLIENT_TYPE = "software.amazon.awssdk.core.SdkClient";
+  private static final String SDK_CLIENT_BUILDER_TYPE = "software.amazon.awssdk.utils.builder.SdkBuilder";
+  private static final String AWS_CLIENT_BUILDER_TYPE = "software.amazon.awssdk.awscore.client.builder.AwsClientBuilder";
+  private static final MethodMatchers BUILD_METHOD = MethodMatchers.create()
+    .ofSubTypes(SDK_CLIENT_BUILDER_TYPE)
+    .names("build")
+    .addWithoutParametersMatcher()
+    .build();
+
+  private static final MethodMatchers REGION_METHOD = MethodMatchers.create()
+    .ofSubTypes(AWS_CLIENT_BUILDER_TYPE)
+    .names("region")
+    .addParametersMatcher("software.amazon.awssdk.regions.Region")
+    .build();
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Collections.singletonList(Tree.Kind.METHOD_INVOCATION);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    MethodInvocationTree invocation = (MethodInvocationTree) tree;
+    if (!BUILD_METHOD.matches(invocation)) {
+      return;
+    }
+    // We first look for a call to region within the same chain of calls.
+    if (setsRegion(invocation)) {
+      return;
+    }
+    // If the call to build is made on a builder variable, we look into initialization and usages for a call to region
+    Optional<VariableTree> declaration = getDeclaration(invocation);
+    if (declaration.isPresent()) {
+      VariableTree actualDeclaration = declaration.get();
+      ExpressionTree initializer = actualDeclaration.initializer();
+      // If the builder variable is not initialized using a method call, we return
+      if (!initializer.is(Tree.Kind.METHOD_INVOCATION)) {
+        return;
+      }
+      MethodInvocationTree initializationChain = (MethodInvocationTree) initializer;
+      if (setsRegion(initializationChain)) {
+        return;
+      }
+      // If no call to region is found in the call, we go to the other usages
+      List<IdentifierTree> usages = actualDeclaration.symbol().usages();
+      // If one of the usages is passing the builder to a method, we assume it might set there
+      if (usages.stream().anyMatch(AwsRegionShouldBeSetExplicitlyCheck::isPassedAsArgument)) {
+        return;
+      }
+      // If the usage sets the region, we return
+      if (usages.stream().anyMatch(AwsRegionShouldBeSetExplicitlyCheck::setsRegion)) {
```

Comment:
We are creating & consuming the same stream twice here - WDYT of having `or`ed boolean expressions in the `anyMatch` only once?

Reply:
Will do